### PR TITLE
Adding the cosmic during collision reconstruction in CMSSW81X

### PR DESCRIPTION
**Pull request by aspiezia**
Implemention of cosmic reconstruction during collision. This could be really usuful for the alignment of the tracker. More details can be found in the following slides, presented at the tracking pog:
https://indico.cern.ch/event/527844/contributions/2160782/attachments/1269896/1881218/AnielloSpiezia_090516.pdf

### CODE REVIEW

**Comment by slava77** [comment]
@aspiezia 
please provide more detailed description (edit the first text block above, at https://github.com/cms-sw/cmssw/pull/14520#issue-155083451)
Please add a link to your presentation in the tracking POG there as well.

**Comment by aspiezia** [comment]
Dear Slava,
I am working to implement cosmic reconstruction during collision. This could be really usuful for the alignment of the tracker. You can find more details in the following slides I have presented at the tracking pog:
https://indico.cern.ch/event/527844/contributions/2160782/attachments/1269896/1881218/AnielloSpiezia_090516.pdf

File: Configuration/StandardSequences/python/ReconstructionCosmics_cff.py [hunk 1]
```diff
@@ -69,6 +69,7 @@
                                              jetsCosmics*
                                              muonsCosmics*
                                              regionalCosmicTracksSeq*
+                                             cosmicCDCTracksSeq*
```
Comment (slava77) [inline Configuration/StandardSequences/python/ReconstructionCosmics_cff.py]: can we merge this into regionalCosmicTracksSeq ?

Comment (aspiezia) [inline Configuration/StandardSequences/python/ReconstructionCosmics_cff.py]: Are you suggesting to to remove the old regionalCosmicTracksSeq? and to call cosmicCDCTracksSeq: regionalCosmicTracksSeq?

File: Configuration/StandardSequences/python/Reconstruction_cff.py [hunk 2]
```diff
@@ -99,9 +102,9 @@
 from FWCore.Modules.logErrorHarvester_cfi import *
 
 # "Export" Section
-reconstruction         = cms.Sequence(localreco*globalreco*highlevelreco*logErrorHarvester)
+reconstruction         = cms.Sequence(localreco*globalreco*highlevelreco*cosmicCDCTracksSeq*logErrorHarvester)
```
Comment (slava77) [inline Configuration/StandardSequences/python/Reconstruction_cff.py]: please remove cosmicCDCTracksSeq from reconstruction\* sequence definitions.
It should be moved to highlevelreco (I don't think it fits well in the globalreco).

**Comment by slava77** [comment]
Hi Aniello,
I happen to know about this PR because I was at the tracking POG meeting.

Please edit the first text block above, at #14520 (comment).
This is for self-documentation.

File: Configuration/StandardSequences/python/ReconstructionCosmics_cff.py (hunk 1 above)
Comment (slava77) [inline Configuration/StandardSequences/python/ReconstructionCosmics_cff.py]: Yes.
Open for discussion though.

File: RecoTracker/SpecialSeedGenerators/python/cosmicCDC_cff.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+hitCollectorForOutInMuonSeeds = TrackingTools.KalmanUpdators.Chi2MeasurementEstimator_cfi.Chi2MeasurementEstimator.clone(
+    ComponentName = cms.string('hitCollectorForOutInMuonSeeds'),
+    MaxChi2 = cms.double(100.0), ## was 30 ## TO BE TUNED
+    nSigma  = cms.double(4.),    ## was 3  ## TO BE TUNED 
+    MaxDisplacement = cms.double(100),
+    MaxSagitta = cms.double(-1.0),
+    MinimalTolerance = cms.double(0.5),
+    appendToDataLabel = cms.string(''),
+)
+cosmicCDCSeeds = RecoTracker.SpecialSeedGenerators.outInSeedsFromStandaloneMuons_cfi.outInSeedsFromStandaloneMuons.clone(
```
Comment (slava77) [inline RecoTracker/SpecialSeedGenerators/python/cosmicCDC_cff.py]: I thought that the first "C" in "CDC" is already for cosmic.
If so "cosmicCDC" is partly tautological
Naming should be somewhat meaningful, non-repetitive.
maybe "cosmic" prefix should be dropped.

**Comment by VinInn** [comment]
@slava77 , my recollection of the discussion in trk-pog meeting was that this sequence can stay in standard reco by default as its has a negligible impact

**Comment by slava77** [comment]
On 5/17/16 1:32 AM, Vincenzo Innocente wrote:
> @slava77 <https://github.com/slava77> , my recollection of the
> discussion in trk-pog meeting was that this sequence can stay in
> standard reco by default as its has a negligible impact
> 

Hi Vincenzo,

my comments so far were in line with that.

I want to avoid a rather specialized sequence name show up in places
where generic/inclusive sequences are supposed to be in order to reduce
configuration support maintenance.
So, if there is an appropriate generic inclusive sequence already
defined as early in the (sub)sequence tree, it should be amended to add
the CDC sequence instead of populating downstream instances with
explicit inclusion of the CDC sequence.

File: Configuration/StandardSequences/python/ReconstructionCosmics_cff.py (hunk 1 above)
Comment (aspiezia) [inline Configuration/StandardSequences/python/ReconstructionCosmics_cff.py]: So, the fact is that here we are using a different approach, not the regional one. So I would keep this name.

File: RecoTracker/SpecialSeedGenerators/python/cosmicCDC_cff.py (hunk 3 above)
Comment (aspiezia) [inline RecoTracker/SpecialSeedGenerators/python/cosmicCDC_cff.py]: I would prefer to keep the "cosmic" in the name, so even if tautological, I would keep this name, if for you is ok.

Comment (slava77) [inline RecoTracker/SpecialSeedGenerators/python/cosmicCDC_cff.py]: how about "cosmicDC" then?

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
It looks like there is a problem in fastsim
e.g. in 135.4:

```
   [2] Calling event method for module CosmicMuonProducer/'cosmicMuons'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for a container with elements of type: TrajectorySeed
Looking for module label: CosmicMuonSeed
```

**Comment by VinInn** [comment]
@lveldere , could you please help with this issue?
thanks

**Comment by lveldere** [comment]
@VinInn 

I'll make a branch with a fix today, possibly before noon.
Lukas

**Comment by lveldere** [comment]
@aspiezia @VinInn 

Find a fixed version of your branch here:
https://github.com/lveldere/cmssw/tree/fastsim-cosmic-fix

Cheers

Lukas

**Comment by lveldere** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mdhildreth** [comment]
+1

File: RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import FWCore.ParameterSet.Config as cms
+
+# seeding
+from RecoMuon.CosmicMuonProducer.cosmicMuons_cfi import *
+import RecoMuon.MuonIdentification.muons1stStep_cfi
+muonsForCosmicDC = RecoMuon.MuonIdentification.muons1stStep_cfi.muons1stStep.clone(
+    inputCollectionLabels = cms.VInputTag("cosmicMuons"),
+    inputCollectionTypes = cms.vstring('outer tracks'),
+    fillIsolation = cms.bool(False),
```
Comment (slava77) [inline RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py]: why are you not using muonsFromCosmic, which are already available in the standard sequence?
They differ from this definition in that they use dt4DCosmicSegments in the muonID variables definition.
So, muonsFromCosmic should give the same set of reco::Muons with the same kinematics, just different ID (matching and timing variables).

If there is no good reason to use this special instance, please switch to muonsFromCosmic.

Comment (aspiezia) [inline RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py]: Using muonsFromCosmic, gives the following error:
----- Begin Fatal Exception 20-May-2016 11:06:12 CEST-----------------------
An exception of category 'MuonDetLayerMeasurements' occurred while
   [0] Processing run: 1 lumi: 1 event: 6
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module CosmicMuonSeedGenerator/'CosmicMuonSeed'
Exception Message:
Cannot get DT RecHits
----- End Fatal Exception -------------------------------------------------

Comment (slava77) [inline RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py]: what did you do exactly and which workflow fails?

**Comment by slava77** [comment]
there is one important element missing: ``` keep recoTrack*_cosmicDCTracks_*_* ``` in FEVT and RECO
(not AOD, if I understood the purpose of this correctly)
to be added in RecoTracker/Configuration/python/RecoTrackerP5_EventContent_cff.py and RecoTracker/Configuration/python/RecoTracker_EventContent_cff.py

**Comment by slava77** [comment]
correcting myself, the following is to be added
```
        'keep recoTracks_cosmicDCTracks_*_*',
        'keep recoTrackExtras_cosmicDCTracks_*_*',
        'keep TrackingRecHitsOwned_cosmicDCTracks_*_*'

```

**Comment by slava77** [comment]
my proposed changes are in the last 3 commits here
https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...slava77:CMSSW_8_1_X_2016-05-18-2300/sign706/pull14520-e28f6eb-testCosMu?expand=1

this avoids using muonsForCosmicDC, which takes up 2/3 of CPU time in the current version of this PR based on tests in 25202 (TTbar with PU35) and 25206 (ZMM with PU35).

**Comment by aspiezia** [comment]
Dear @slava77, thanks for the help on the muons from cosmic collection.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import FWCore.ParameterSet.Config as cms
+
+# seeding
+import RecoTracker.SpecialSeedGenerators.outInSeedsFromStandaloneMuons_cfi
+import TrackingTools.KalmanUpdators.Chi2MeasurementEstimator_cfi 
+hitCollectorForOutInMuonSeeds = TrackingTools.KalmanUpdators.Chi2MeasurementEstimator_cfi.Chi2MeasurementEstimator.clone(
```
Comment (slava77) [inline RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py]: this has to be renamed to something else, e.g. hitCollectorForCosmicDCMuonSeeds
(the  ComponentName will need to be updated as well as hitCollector inside of the cosmicDCSeeds below.
Current version overrides the definition from RecoTracker/SpecialSeedGenerators/python/outInSeedsFromStandaloneMuons_cfi.py

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14520	9773421
- code changes are in line with the description (and a follow up review): added a sequence leading to cosmicDCTracks (saved in FEVT and RECO), based on out-in seeding from cosmicMuons (similar setup to the out-in muon tracking iteration)
- jenkins tests pass and comparisons with baseline show no difference (monitored objects are not affected)
- local tests with ZMM PU35 show reasonably small CPU time spent in the new sequence, under 0.1% of full reco (~15 s/evt in the time units of the test)
```
         9.58 ms/ev cosmicDCSeeds
         1.88 ms/ev cosmicDCCkfTrackCandidates
        0.64 ms/ev cosmicDCTracks
```

**Pull request merged**